Add `gh repo edit --enable-discussions`

azu opened the pull request:
This PR add `--enable-discussions` flag to `gh repo edit` command.

## Test plan

```bash
$ make 
$ ./bin/gh repo edit --help
Edit repository settings.

To toggle a setting off, use the `--flag=false` syntax.


USAGE
  gh repo edit [<repository>] [flags]

FLAGS
      --add-topic strings        Add repository topic
      --allow-forking            Allow forking of an organization repository
      --allow-update-branch      Allow a pull request head branch that is behind its base branch to be updated
      --default-branch name      Set the default branch name for the repository
      --delete-branch-on-merge   Delete head branch when pull requests are merged
  -d, --description string       Description of the repository
      --enable-auto-merge        Enable auto-merge functionality
      --enable-discussions       Enable discussions in the repository
      --enable-issues            Enable issues in the repository
      --enable-merge-commit      Enable merging pull requests via merge commit
      --enable-projects          Enable projects in the repository
      --enable-rebase-merge      Enable merging pull requests via rebase
      --enable-squash-merge      Enable merging pull requests via squashed commit
      --enable-wiki              Enable wiki in the repository
  -h, --homepage URL             Repository home page URL
      --remove-topic strings     Remove repository topic
      --template                 Make the repository available as a template repository
      --visibility string        Change the visibility of the repository to {public,private,internal}

INHERITED FLAGS
  --help   Show help for command

ARGUMENTS
  A repository can be supplied as an argument in any of the following formats:
  - "OWNER/REPO"
  - by URL, e.g. "https://github.com/OWNER/REPO"

EXAMPLES
  # enable issues and wiki
  gh repo edit --enable-issues --enable-wiki

  # disable projects
  gh repo edit --enable-projects=false

LEARN MORE
  Use 'gh <command> <subcommand> --help' for more information about a command.
  Read the manual at https://cli.github.com/manual

# test some repositories
$ ./bin/gh repo edit azu/github-advisory-database-rss --enable-discussions
✓ Edited repository azu/github-advisory-database-rss
$ ./bin/gh repo edit azu/hubmemo --enable-discussions --enable-auto-merge --delete-branch-on-merge
✓ Edited repository azu/hubmemo
```


fix #6896

azu committed:
Add gh repo edit --enable-discussions

mislav commented on pkg/cmd/repo/edit/edit.go:
```diff
@@ -181,6 +184,7 @@ func editRun(ctx context.Context, opts *EditOptions) error {
 			"hasIssuesEnabled",
 			"hasProjectsEnabled",
 			"hasWikiEnabled",
+			"hasDiscussionsEnabled",
```
We cannot unconditionally query the `hasDiscussionsEnabled` GraphQL field because it's a fairly recent API that does not exist on older GitHub Enterprise versions. (GitHub CLI is also compatible with GitHub instances that are on hosts other than `github.com`.)

There is some prior art to conditionally querying GraphQL fields, but I'd say that for now it's the easiest to avoid querying this value and instead default to `false` in interactive flow (which is the only place that otherwise uses this value, I think?)

azu replied:
I will comment out it.

```diff
diff --git a/pkg/cmd/repo/edit/edit.go b/pkg/cmd/repo/edit/edit.go
index 01bdead1..ae670bcf 100644
--- a/pkg/cmd/repo/edit/edit.go
+++ b/pkg/cmd/repo/edit/edit.go
@@ -184,7 +184,8 @@ func editRun(ctx context.Context, opts *EditOptions) error {
 			"hasIssuesEnabled",
 			"hasProjectsEnabled",
 			"hasWikiEnabled",
-			"hasDiscussionsEnabled",
+			// TODO: GitHub Enterprise Server does not support has_discussions yet
+			// "hasDiscussionsEnabled",
 			"homepageUrl",
 			"isInOrganization",
 			"isTemplate",
@@ -279,7 +280,8 @@ func interactiveChoice(r *api.Repository) ([]string, error) {
 		optionIssues,
 		optionMergeOptions,
 		optionProjects,
-		optionDiscussions,
+		// TODO: GitHub Enterprise Server does not support has_discussions yet
+		// optionDiscussions,
 		optionTemplateRepo,
 		optionTopics,
 		optionVisibility,
```

Should I hide "[ ] Discussions" too?

Before

<img width="654" alt="image" src="https://user-images.githubusercontent.com/19714/214171007-cdf5ee67-4da1-4e61-9558-0da96dfec801.png">

After
<img width="610" alt="image" src="https://user-images.githubusercontent.com/19714/214170765-72d6352c-5c7f-4611-b8a4-b716adccd022.png">

azu replied:
https://github.com/cli/cli/pull/6903/commits/0877f56e7449544781fa88f9b6c9b85ebc88f355

mislav reviewed (changes requested):
Thanks for the fast PR. Looks great, but we'd have to be mindful of GitHub Enterprise Server compatibility

azu committed:
hide Discussions options in interactive mode

Avoid to  `hasDiscussions` query to GitHub Enterprise Server.
GHE does not support `hasDiscussions` query yet.

mislav committed:
Add hasDiscussionsEnabled to JSON fields for repos

mislav reviewed (approved):
Fantastic. Thank you! 🙇
